sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,591 @@
+    The three generalized coordinates (thetax, thetay and thetax) are always
+    the three angles along three perpendicular vectors and the three
+    generalized speeds (omegax, omegay and omegaz) are the angular velocities
+    along those vectors respectively.
+
+    Parameters
+    ----------
+    name: String
+        Name of the joint which makes it unique. Should be different from other
+        joints.
+    parent: Body
+        Instance of Body class serving as the parent in the joint.
+    child: Body
+        Instance of Body class serving as the child in the joint.
+    parent_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the parent will be connected to child.
+        3 Tuple defines the values of x, y and z directions w.r.t parent's
+        frame. If it is not supplied, center of mass is added as default.
+    child_point_pos: 3 Tuple, Optional
+        Defines the joint's point where the child will be connected to parent.
+        3 Tuple defines the values of x, y and z directions w.r.t child's frame.
+        If it is not supplied, center of mass is added as default.
+    parent_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with child's
+        axis before adding joint. If it is not passed, default is x axis in
+        parent's frame.
+    child_axis: Vector, Optional
+        Defines the orientation as a vector which must be aligned with parent's
+        axis before adding joint. If it is not passed, default is x axis in
+        child's frame.
+
+    Example
+    -------
+    Adds spherical Joint between parent's masscenter and a point at unit distance
+    along child's y axis.
+    >>> from sympy.physics.mechanics import Body, SphericalJoint
+    >>> parent = Body('parent')
+    >>> child = Body('child')
+    >>> spherical_joint = SphericalJoint('sphericaljoint', parent, child, \
+                                         child_point_pos=(0, 1, 0)
+    >>> spherical_joint.coordinates
+    [thetax(t), thetay(t), thetaz(t)]
+    >>> spherical_joint.speeds
+    [omegax(t), omegay(t), omegaz(t)]
+
+    """
+    def __init__(self, name, parent, child, parent_point_pos=None,
+                 child_point_pos=None, parent_axis=None, child_axis=None):
+
+        super(SphericalJoint, self).__init__(name, parent, child,
+                                             parent_point_pos, child_point_pos,
+                                             parent_axis, child_axis)
+
+    def apply_joint(self):
+        thetax = dynamicsymbols('thetax')
+        thetay = dynamicsymbols('thetay')
+        thetaz = dynamicsymbols('thetaz')
+        thetaxd = dynamicsymbols('thetax', 1)
+        thetayd = dynamicsymbols('thetay', 1)
+        thetazd = dynamicsymbols('thetaz', 1)
+        omegax = dynamicsymbols('omegax')
+        omegay = dynamicsymbols('omegay')
+        omegaz = dynamicsymbols('omegaz')
```

Comment:
These names should be unique wrt to each joint. If I have 5 spherical joints in a model, then you don't end up with unique coordinates and speeds.